### PR TITLE
feat(backend): add window size api

### PR DESCRIPTION
**Pull request by benjajaja**
Depends on https://github.com/crossterm-rs/crossterm/pull/790, and should update crossterm in an independent PR. Related to #215

For image (sixel, iTerm2, Kitty...) support that handles graphics in
terms of `Rect` so that the image area can be included in layouts.

For example: an image is loaded with a known pixel-size, and drawn, but
the image protocol has no mechanism of knowing the actual cell/character
area that been drawn on. It is then impossible to skip overdrawing the
area.

Returning the window size in pixel-width / pixel-height, together with
colums / rows, it can be possible to account the pixel size of each cell
/ character, and then known the `Rect` of a given image, and also resize
the image so that it fits exactly in a `Rect` and avoid artifacts.

image

### CODE REVIEW

File: Cargo.toml
Comment (joshka) [inline Cargo.toml:41]: I think I'd prefer to wait until this is merged to merge this into ratatui. This is fine as a draft for now however.

You can also set this in a patch section and leave the main dependency in place. See https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section

```toml
[patch.crates-io]
crossterm = { git = "https://github.com/benjajaja/crossterm", rev = "ad84a3829d55eac27cfdb7791b2e34c1732040ad", optional = true }
```

File: examples/window_size.rs
Comment (joshka) [inline examples/window_size.rs:1]: It could be worthwhile adding some doc comments to help a reader understand what this example shows.

File: src/backend/crossterm.rs
Comment (joshka) [inline src/backend/crossterm.rs:89]: I see this is removed a bunch of times. Can you please find out or explain why this was necessary in the first place?

File: src/backend/mod.rs
Comment (joshka) [inline src/backend/mod.rs:117]: Do you think it makes sense to split this to two methods with more explicit names instead of a single returning a tuple? Is the Rect ever not positioned at x=0, y=0?

File: src/backend/termwiz.rs
Comment (joshka) [inline src/backend/termwiz.rs:237]: Replace with `u16::try_from(i).unwrap_or(u16::MAX);` (probably obvious enough to just inline into the methods that use it)

File: src/backend/test.rs
Comment (joshka) [inline src/backend/test.rs:181]: What's the significance of 7 and 12 here? Can you document it?

**Review by joshka — CHANGES_REQUESTED** [review]
Rewrite the commit message title in the convention commit style using imperative tone (e.g. `feat(backend): add window_size_in_pixels() method`).

This is pretty close to being mergable (or will be once the crossterm PR is merged), so can you please rebase and squash your commits into a single commit and setup commit signing? See https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits for info.

**Comment by joshka** [comment]
Backend is woefully under-tested. I wonder if there's some way to fix that.

File: Cargo.toml
Comment (benjajaja) [reply Cargo.toml:41]: Oooh TIL, that's so cool, thank you!

File: examples/window_size.rs
Comment (benjajaja) [reply examples/window_size.rs]: I removed this in the end, as it's not a valuable demo/example.

File: src/backend/crossterm.rs
Comment (benjajaja) [reply src/backend/crossterm.rs:91]: It seems like the `queue` macro, or `QueueableCommand::queue`, now return `io::Result` making the `map_error` unnecessary.

File: src/backend/mod.rs
Comment (benjajaja) [reply src/backend/mod.rs]: I say no, because 2 out of 3 implementations return both things at once in one call, due to the underlying API.

File: src/backend/termwiz.rs
Comment (benjajaja) [reply src/backend/termwiz.rs]: Much better, thank you!

File: src/backend/test.rs
Comment (benjajaja) [reply src/backend/test.rs]: It was from when it was `font_size`, not `window_size` - my bad.
Updated to a better value and added a comment.

File: src/backend/mod.rs
Comment (benjajaja) [reply src/backend/mod.rs]: Oh and the `Rect` is just like `size`, always x=0,y=0.

**Comment by benjajaja** [comment]
Regarding tests, I don't think it's really feasible to cover this. `window_size()` really depends on being run as a graphical application, i.e. xorg or wayland.

**Review by joshka — CHANGES_REQUESTED** [review]
I think there was a mention of this not yet working with windows on crossterm if I read the linked issue right. Can you make sure to document this in the crossterm docs for the window_size method?

File: src/backend/mod.rs
Comment (joshka) [reply src/backend/mod.rs]: I've been thinking about this a little - I think there's a few places within various widgets where we care more about the particular offset that a size. A good example of that `Buffer::empty()` which takes a Rect, but really only needs that to be a Size. We use this a lot for tests. So we might treat a Rect as a Point(x,y) and a Size(width, height).

It's probably out of scope for this PR to do all that however. PixelSize is still a bit ambiguous (I could see users getting confused over whether it is the size of a cell in pixels or the size of a window?)

What about:
```rust
struct WindowSize {
  size_in_rows_and_cols: Size,
  size_in_pixels: Size,
}
```

> Changing the return type of size() would also mean that this is a breaking change. Perhaps it would be better to have window_size() just return (u16, u16) as you mentioned? What if that was named `size_in_pixels()`?

**Comment by joshka** [comment]
BTW, another PR upgraded crossterm 0.27 and brought in the changes to the error bits. Can you rebase this on main please?

File: src/terminal.rs
Comment (benjajaja) [inline src/terminal.rs:423]: `last_known_size` was a `Rect` that was only set from `with_options()` and `resize()`. Both did only have non-zero x,y for `Viewport::Fixed`. This method only runs for `Viewport::Inline(_)` and it is the only method that uses the field.

**Comment by benjajaja** [comment]
I tried adding some test coverage for `insert_before`, but that did require to implement `clear_region()` and `append_line()` for `TestBackend` for which I don't know exactly how to implement the different "clear types", e.g. `AfterCursor` etc. In general, it seemed a bit pointless to test `insert_before` against `TestBackend`.

File: src/backend/mod.rs
Comment (joshka) [inline src/backend/mod.rs:118]: How about leaving this as Rect and just updating the window_size method?

otherwise nit: doc comment => Size instead of Rect

File: src/layout.rs
Comment (joshka) [inline src/layout.rs:599]: Do we actually need this truncation for anything?

**Review by joshka — CHANGES_REQUESTED** [review]
I think perhaps make this non-breaking with the existing size() and resize() continuing to take Rects? We can come back to that decision at a later date however.

Aiming for a 0.23.0 release after the weekend, so this should be in that if you can make the changes in time.

**Comment by benjajaja** [comment]
@joshka I toned this down to only adding the new `window_resize()` method, still introducing the new `Size` struct but not using it in src/terminal.rs so no breaking changes. Also dropped the truncation in `Size` copied from `Rect`, as its main purpose is to keep `area` (and maybe indirectly `Buffer::content`) in bounds, and there are no more conversions between `Size` and `Rect` now.

It's cool if we can get this into the release, but it's not critical like the cell skipping, as the window pixel size can be queried outside of `ratatui`.

--------------
Some thoughts on `Size` vs `Rect` in `Terminal`:

In `Terminal`, there is a `last_known_size` that may have non-zero x,y if the `viewport` is `Viewport::Fixed(Rect)`. However it seems like the field is used exclusively by `Viewport::Inline(u16)`.

In general, I think it could be healthy to cut down a few public methods in `Terminal`: `autoresize`, `resize`, `size` (is just a wrapper to `backend.size()`), and `area` of `CompletedFrame` returned by `draw`.

File: src/backend/mod.rs
Comment (joshka) [inline src/backend/mod.rs:65]: nit: what do you think about naming this just `pixels` given that the struct is named **`Window`**`Size`?

**Review by joshka — APPROVED** [review]
This looks good now. Thanks again. This seems like something we could merge before release. I'll flip it to the other maintainers to take a look.

Comments are non-blocking

File: src/backend/mod.rs
Comment (kdheepak) [reply src/backend/mod.rs:65]: I think there should be a doc string here too, right? Maybe describing when someone might want to use this.

**Review by mindoodoo — APPROVED** [review]
Looks good ! Thanks for this !

I'll have to take a look at the image crate soon ! Exciting stuff.

@joshka what do you think about the scope of the conventional commit PR title ? let me know if anything else would be more suited. Thanks for all the review on this !

**Comment by joshka** [comment]
> @joshka what do you think about the scope of the conventional commit PR title ? let me know if anything else would be more suited. Thanks for all the review on this !

I think it's probably ok as is. Merging this now. Thanks again @benjajaja. This is going to lead to some fun apps.

**Pull request merged**
